akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java

```diff
@@ -0,0 +1,338 @@
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    for (int i = 0; i < depth; ++i) {
+      table[i][buckets[i]] += count;
+    }
+    size += count;
+  }
+
+  /**
+   * Similar to {@code add}, however we reuse the fact that the hask buckets have to be calculated for {@code add}
+   * already, and a separate {@code estimateCount} operation would have to calculate them again, so we do it all in one go.
+   */
+  public long addAndEstimateCount(String item, long count) {
+    if (count < 0) {
+      // Actually for negative increments we'll need to use the median
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    for (int i = 0; i < depth; ++i) {
+      table[i][buckets[i]] += count;
+    }
+    size += count;
+    return estimateCount(buckets);
+  }
+
+  public long size() {
+    return size;
+  }
+  
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   */
+  public long estimateCount(long item) {
+    long res = Long.MAX_VALUE;
+    for (int i = 0; i < depth; ++i) {
+      res = Math.min(res, table[i][hash(item, i)]);
+    }
+    return res;
+  }
+
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   */
+  public long estimateCount(String item) {
+    int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    return estimateCount(buckets);
+  }
+
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   * 
+   * @param buckets the "indexes" of buckets from which we want to calculate the count
+   */
+  private long estimateCount(int[] buckets) {
+    long res = Long.MAX_VALUE;
+    for (int i = 0; i < depth; ++i) {
+      res = Math.min(res, table[i][buckets[i]]);
+    }
+    return res;
+  }
+  
+  /**
+   * This is copied from https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/hash/MurmurHash.java
+   * Which is Apache V2 licensed.
+   * <p></p>
+   * This is a very fast, non-cryptographic hash suitable for general hash-based
+   * lookup. See http://murmurhash.googlepages.com/ for more details.
+   * <p/>
+   * <p>
+   * The C version of MurmurHash 2.0 found at that site was ported to Java by
+   * Andrzej Bialecki (ab at getopt org).
+   * </p>
+   */
+  private static class MurmurHash {
```

Comment:
I know something better: **Use CRC32!** First, it is a reusable mutable construct which takes an array instead of allocating one, second, since Java8 **it is a compiler intrinsic** compiling down to CPU supported CRC instructions! Tops, CRC32 is still a bloody good hash function. You can try with turning the intrinsics off (-XX:-UseCLMUL -XX:-UseCRC32Intrinsics) and on (I think that's the default).